### PR TITLE
Adding Go coding style guidelines

### PR DESCRIPTION
**Pull request by stevvooe**
In the spirt of #14756, I've assembled some lightweight coding style guidelines. Let's keep controversy to a minimum.

cc @icecrime @thaJeztah 

Signed-off-by: Stephen J Day stephen.day@docker.com

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -385,3 +385,48 @@ do need a fair way to deal with people who are making our community suck.
...
+
+Unless explicitly stated, we follow all coding guidelines from the Go
+community. While some of these standards may seem arbitrary, they somehow seem
+to result in a solid, consistent codebase.
+
+It is possible that the code base does not currently comply with these
+guidelines. We are not looking for a massive PR that fixes this, since that
+goes against the spirit of the guidelines. All new contributions should make a
+best effort to clean up and make the code base better than they left it.
+Obviously, apply your best judgement. Remember, the goal here to make the code
```
Comment (thaJeztah) [inline CONTRIBUTING.md]: s/the goal here to/the goal here is to/

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -385,3 +385,48 @@ do need a fair way to deal with people who are making our community suck.
...
+   noALongVariableNameLikeThisIsNotMoreClearWhenASimpleCommentWouldDo. In
+   practice, short methods will have short variable names and globals will
+   have longer names.
+7. No underscores in package names. If you need a compound name, step back,
+   and re-examine why you need a compound name. If you still think you need a
+   compound name, lose the underscore.
+8. No utils or helpers packages. If a function is not general enough to
+   warrant it's own package, it has not been written generally enough to be a
+   part of a util package. Just leave it unexported and well-documented.
+9. All tests should run with `go test` and outside tooling should not be
```
Comment (runcom) [inline CONTRIBUTING.md]: can we go deeper here and mention something like "docker is using `gocheck`", "docker command should be run with `dockerCmd*` family of commads" etc etc?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -385,3 +385,48 @@ do need a fair way to deal with people who are making our community suck.
...
+base easier for people to navigate and understand. Always keep that in mind
+when nudging others to comply.
+
+The rules:
+
+1. All code should be formatted with `gofmt -s`.
+2. All code should pass the default levels of
+   [`golint`](https://github.com/golang/lint).
+3. All code should follow the guidelines covered at
+   https://github.com/golang/go/wiki/CodeReviewComments.
```
Comment (thaJeztah) [inline CONTRIBUTING.md]: can you change this to

`[Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments]`?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -385,3 +385,48 @@ do need a fair way to deal with people who are making our community suck.
...
+4. Comment the code. Tell us the why, the history and the context.
+5. Document _all_ declarations and methods, even private ones. Declare
+   expectations, caveats and anything else that may be important. If a type
+   gets exported, having the comments already there will ensure it's ready.
+6. Variable name length should be proportional to it's context and no longer.
+   noALongVariableNameLikeThisIsNotMoreClearWhenASimpleCommentWouldDo. In
```
Comment (thaJeztah) [inline CONTRIBUTING.md]: can you put the extremelylongvariablename in ticks?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -385,3 +385,48 @@ do need a fair way to deal with people who are making our community suck.
...
+   warrant it's own package, it has not been written generally enough to be a
+   part of a util package. Just leave it unexported and well-documented.
+9. All tests should run with `go test` and outside tooling should not be
+   required. No, we don't need another unit testing framework. Assertion
+   packages are acceptable if they provide _real_ incremental value.
+10. Even though we call these "rules" above, they are actually just
+    guidelines. Since you've read all the rules, you now know that.
+
+If you are having trouble getting into the mood of idiomatic Go, we recommend
+reading through [`Effective Go`](http://golang.org/doc/effective_go.html). The
```
Comment (thaJeztah) [inline CONTRIBUTING.md]: drop the ticks around "Effective Go"

File: CONTRIBUTING.md (hunk 2 above)
Comment (thaJeztah) [inline CONTRIBUTING.md]: I don't think we should go too much into depth here. We could add a more in-depth article as a follow up? This is just to get started

**Comment by thaJeztah** [comment]
Me likes! Added some (mainly nits) while reading, in addition;

- this needs a link under "Topics"
- there's some overlap with the [conventions](https://github.com/stevvooe/docker/blob/coding-style-guidelines/CONTRIBUTING.md#conventions) section. No problem I think, but you could add a link to the coding style section from there.

I like what you've written. Short, not too much in depth, but enough to get a rough idea. I'm thinking if we should include this in this document, or move it to a separate file ('CODING-STYLE.md` / `CODING-CONVENTIONS.md`?). Perhaps this could be reused in other repositories as well.

File: CONTRIBUTING.md (hunk 2 above)
Comment (runcom) [inline CONTRIBUTING.md]: mmm alright I do see this as the HACKING.rst file in openstack nova repo and they're actually mentioning tips like the ones I mentioned, not biggie tho :)

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -385,3 +385,48 @@ do need a fair way to deal with people who are making our community suck.
...
+5. Document _all_ declarations and methods, even private ones. Declare
+   expectations, caveats and anything else that may be important. If a type
+   gets exported, having the comments already there will ensure it's ready.
+6. Variable name length should be proportional to it's context and no longer.
+   noALongVariableNameLikeThisIsNotMoreClearWhenASimpleCommentWouldDo. In
+   practice, short methods will have short variable names and globals will
+   have longer names.
+7. No underscores in package names. If you need a compound name, step back,
+   and re-examine why you need a compound name. If you still think you need a
+   compound name, lose the underscore.
```
Comment (duglin) [inline CONTRIBUTING.md]: Want to mention to use camelCase?

**Comment by duglin** [comment]
overall LGTM

File: CONTRIBUTING.md (hunk 2 above)
Comment (thaJeztah) [inline CONTRIBUTING.md]: Oh, I think it's useful information, but it might distract from the "core" information here (I think)

Comment (stevvooe) [inline CONTRIBUTING.md]: @runcom These are meant to be brief guidelines. If there are particular choices, they are outside the scope of the style guidelines.

If I had a real opinion, we would not use anything outside the standard library and a small `CheckX` package defined internally.

Comment (runcom) [inline CONTRIBUTING.md]: @stevvooe @thaJeztah :+1:

File: CONTRIBUTING.md (hunk 4 above)
Comment (stevvooe) [inline CONTRIBUTING.md]: I added the missing "comma", as well. ;)

Comment (thaJeztah) [inline CONTRIBUTING.md]: Yeah! Uber-nit-pick.

File: CONTRIBUTING.md (hunk 6 above)
Comment (stevvooe) [inline CONTRIBUTING.md]: This is covered by golint.

**Comment by stevvooe** [comment]
@thaJeztah I would always expect this kind of thing to be in `CONTRIBUTING.md`, although we can put wherever makes sense. For this PR, this location is probably okay.

Do we actually want this as part of the "Conventions" section?

**Comment by thaJeztah** [comment]
@stevvooe 

> For this PR, this location is probably okay.

sgtm

> Do we actually want this as part of the "Conventions" section?

Just a mention/link to this section would do I think. This document is getting quite lengthy, and knowing some people are lazy, anything that makes it more likely to read this would help. Just making it easier for them to find.

**Comment by stevvooe** [comment]
@thaJeztah Updated for the nits.

Modified Rule 3 to include _Effective Go_.

**Comment by stevvooe** [comment]
@thaJeztah Added.

**Comment by thaJeztah** [comment]
LGTM, thanks!

Think this needs (one) more maintainers LGTM because this is not strictly "docs".

ping @icecrime for approval

**Comment by jfrazelle** [comment]
LGTM

**Comment by thaJeztah** [comment]
I don't see objections here, so I'll merge. Thanks @stevvooe !

**Pull request merged**